### ISSUE TITLE
Browser Issues: ITP, Cookies, Redirection

### ISSUE THREAD

**Issue by osiset** [osiset/laravel-shopify#522]
**Feature**: Currently there is no support for ITP cookies, theres also lingering issues with cookies outside of my control sadly.

I am not sure if there will be a clean solution for this as adding ITP support really bloats the flow and quite honestly is frustrating to deal with the baggage Chrome and Chrome-like browsers bring with their isolated standards.

I am not sure when I can complete this, or what it will take, or even if I can do it alone. However, theres been a growing number of cookie issues (again, outside of my control), so I will attempt this as the next feature!

Any opened cookie-related issues will be closed and this issue will be referenced. I will pool your reports and use them to attempt a fix.

Thanks!

**Comment by osiset** [comment]
@Kyon147 FYI. If you see any issues which may regard cookies, feel free to close and reference this issue as master :) going to pool the reports and find a way to fix ITP.

**Comment by Kyon147** [comment]
> @Kyon147 FYI. If you see any issues which may regard cookies, feel free to close and reference this issue as master :) going to pool the reports and find a way to fix ITP.

No problem, also if you need some help with the issue - I'll pull a fork of the repo and try and take a look as well. As you say you are one person, so the more support the better for you and the package as a whole. If you'd like me to :)

**Comment by osiset** [comment]
Any help is welcome. I had a branch last year where I did some minor ITP work, so I may revive this and massage it into the current code base and see where I can get with safari.

**Comment by Kyon147** [comment]
Okay, I'll wait a little bit to see what you do with the existing work
before jumping into it.

That way don't risk doing the same stuff etc

**Comment by osiset** [comment]
Following up, as a current plan of attack.

I had previous code started to help deal with ITP which showed promise. The code is no longer compatible however, but I will pick out the bits.

I will be attempting to debug this issue slowly with Safari/Chrome and try and implement ITP support as I go.

**Comment by osiset** [comment]
Is anyone aware of a way to replicate this without Safari?

Firefox and Chrome I have both set to block third party cookies, but its still allowing me through.

**Comment by darrynten** [comment]
We've started getting 419s on some calls (but not others) - not sure if it's related though and not seeing anything in the logs.

Am investigating, will advise.

**Comment by faisal-saeed-brainx** [comment]
Hi @osiset, I am getting the samesite cookie issue in chrome incognito.

**Comment by pagameba** [comment]
Hi @osiset, I am also getting this issue with Chrome Version 83.0.4103.116 (Official Build) (64-bit) setting up a brand new laravel app and this package.  In Safari, it works without the cookie error.  Not sure if it is a different issue or related to this, after I install the app on shopify, it asks me to login to the store on the app side and just continually redirects back to the login page rather than redirecting to shopify.  If I visit the app through the store admin, it seems to work just fine.  At least once, it did start working but I cannot reproduce the chain of events that got me to that point.

**Comment by osiset** [comment]
@pagameba Can you please link me to a pastebin or something of your current config/session.php? I will copy it and give it a try on latest Chrome.

**Comment by pagameba** [comment]
session.php contents: https://pastebin.com/3jiVhpXL

Totally willing to try anything or provide any info, just let me know

**Comment by osiset** [comment]
@pagameba Sadly not able to repeat with my Docker setup. Will try a full deployment to my Heroku test environment though.

How are you accessing the app? ngrok, docker/https, ... ?

**Comment by pagameba** [comment]
edited my comment above with the details, I'm using ngrok/docker/https - willing to help out however I can, I'm going to spend some time today trying to understand how the flow of control works.  I'm not a laravel expert so I get easily lost in some of the magic :)  I had someone on my team create a node version of the app and that does work with Chrome so I assume it is possible to get it working here too and would prefer a PHP solution as its a better fit for our stack.

**Comment by pagameba** [comment]
I added some debug logging.

The `checkSameSiteNoneCompatible()` method in `CookieHelper.php` is returning `true`.

The redirect back to login happens in `AuthController.php` because `$status` is null returning from `AuthenticateShop::__invoke()` because the call to `$this->apiHelper->verifyRequest` is returning false.

The `Request` object seems valid in that it has a shop, a timestamp and an hmac (plus a code which doesn't seem to be used in verifyRequest).  However, the params also contain `"q":"/authenticate"` ... removing that from the params (           `unset($params['q']);`) before computing the hmac hash allows it to work correctly and redirect to the shopify store after authentication.

After redirection, I see the app loaded correctly in the shopify admin interface. 

![image](https://user-images.githubusercontent.com/249841/88070737-0414b480-cb41-11ea-84f1-86a4e8a27fc7.png)

 However, in the Apps list, I still see this after a fresh install of the app.

![image](https://user-images.githubusercontent.com/249841/88070797-155dc100-cb41-11ea-9195-14333a4796a5.png)

I would recommend changing `verifyRequest` to only include keys from `$params` that should be in the hmac hash calculation, I think that is part of `osiset/BasicShopifyAPI` though.

**Comment by osiset** [comment]
@pagameba Great digging! Shopify keeps adding stuff to the URLs the past few years, I can't tell anyone what they're including in the HMAC signature. When I originally wrote the verify method, the only things existing in the URL was hmac/shop/timestamp.

I've merged in your code and will push out a patch fix hopefully tonight. While this won't fix the ITP issue, it should fix the redirect to login by looks of it, which is great.

**Comment by pagameba** [comment]
:+1: I didn't get any further with the cookie issue other than to determine that it does seem to see that the browser is SameSite/None compatible and that it seems to set appropriate values for the cookie based on that.  Totally willing to dig further just not quite sure where to go next :)

**Comment by osiset** [comment]
@pagameba Actually, may be an issue with this. Shopify's docs are noting it uses state, locale, and possibly nonce.

It may be safer to check and remove the "q" part.

I've tried several apps and can't see this "q" variable, where did you notice it?

**Comment by osiset** [comment]
@pagameba Yup and protocol by looks of it. If you can add those in and send a PR, I'll re-test again.

I tested with the original PR and noticed a test failed on this package, looking into the test, its an HMAC from Shopify's docs, which lead me to see it does use things like state/locale/nonce/protocol.

Edit: Its possible this is your issue with the "q": https://laracasts.com/discuss/channels/general-discussion/laravel-request-q-parameter. Nginx try_files would need an update.

**Comment by pagameba** [comment]
Ah, I think you are correct about the q issue, I will fix that in my container and try again. I'm looking at the [shopify documentation](https://shopify.dev/tutorials/authenticate-with-oauth#verification) and I don't see an explicit list of parameters used in the hmac calculation.  Perhaps just removing `q` would be sufficient since it seems I am not the only one with this setup - pretty sure its just a boilerplate configuration I copied from somewhere?  I'm happy to add a PR for the other parameters if you think that's the right way to go though.

**Comment by osiset** [comment]
Given it's an nginx issue, I wouldn't want to remove it in the library. Hmm
________________________________

**Comment by darrynten** [comment]
> Any help is welcome. I had a branch last year where I did some minor ITP work, so I may revive this and massage it into the current code base and see where I can get with safari.

Which branch?

**Comment by yrk2yrk** [comment]
Having issue with chrome incognito and safari private mode. On chrome it says: Call to a member function api() on null.
On safari it just takes again to login page. I checked the auth values but don't see 'q' parameter in it. 

I am using nginx as reverse proxy server. It doesn't have try_files option configured. I tried adding the updated try_files option but it didn't work.

**Comment by osiset** [comment]
Unable to get this to trigger in Chrome Incognito. I don't have a Mac... I'll have to hunt one down to loan and try that way. Will have to hold this until then, nothing I can reproduce without Safari it seems.

**Comment by yrk2yrk** [comment]
What is the version of your Chrome browser? For me, it is failing on 84.0 but works on 79.0. I tried both in incognito.

**Comment by osiset** [comment]
Version 84.0.4147.89 (Official Build) (64-bit) :(

**Comment by yrk2yrk** [comment]
Okay, I will just test this more and share details. Could be OS specific.

**Comment by yrk2yrk** [comment]
Update: The app works fine with normal Chrome mode. But it cannot access session/cookie after auth redirect in Incognito mode. If Incognito setting is set to "Block all 3rd party cookies" which is default I guess, it doesn't work.  If I change it to "Allow all cookies" then it works. Also, if I don't do redirect through app bridge, it works on my domain. 

When my app code tries to access shop.json after redirect it gives the error. But I could access the shop JSON data using debug mode as mentioned [here](https://github.com/osiset/laravel-shopify/wiki/Usage) after auth redirect. 

I am testing on Chrome for Mac OS.

**Comment by osiset** [comment]
@yrk2yrk I reinstalled my local Laradock setup, same issue - even with 3rd party cookies blocked it still lets me in. Are you doing this externally or locally? I may try to redeploy this to Heroku again to see if maybe the local is the issue.

**Comment by yrk2yrk** [comment]
I am doing it externally. My app is hosted on AWS. I don't know if this is package related or something else. But I tried every possible option. Can I email you my app link to test?

**Comment by osiset** [comment]
@yrk2yrk Please yes, that would help. Just need to determine if my browsers are the issue or its simple because its locally hosted via Docker through a .local domain that the cookies work.

**Comment by yrk2yrk** [comment]
Just emailed you the link.

**Comment by osiset** [comment]
Thanks. I get the "Trying to get property 'name' of non-object on null" error, I see cookies in Chrome, but cookies in the debug are blank. Hmm, maybe it is because its local. I will give it a try on a deployed location and go from there.

**Comment by yrk2yrk** [comment]
Yes, I was trying different config options for App bridge redirect so getting a different error now. Let me know if it works for you on live deployment.

**Comment by GuiguiWeb** [comment]
@osiset Cookieless App Bridge auth beta now available ! [developer documentation](https://shopify.dev/tools/app-bridge/authentication)

**Comment by osiset** [comment]
I've seen this yes. However it seems geared towards SPAs, which everyone is not using. It doesn't exactly solve our issue of Laravel using cookies to store the information.

I'm going to repush my local to Heroku and see if I can trigger the ITP issue.

**Comment by osiset** [comment]
I must be bad luck. I setup a brand new Laravel install, deployed it.

For both Firefox and Incognito Chrome.. theres no cookies but auth is fine. Chrome normal, cookies do show.

I'm curious how Laravel is saying I'm logged in but theres no session cookies being used. I'll have to dig into the internal code, but I'm wondering if its falling back to file-based cookies? Really odd. Any insight from anyone would be helpful too.

![image](https://user-images.githubusercontent.com/2420282/89037957-0681c600-d31a-11ea-88ac-d6f5cc0511da.png)

**Comment by osiset** [comment]
I have a loan of a Macbook for a little bit and will be able to check this out now with Safari. Still unsure how well this will fair.

**Comment by faisal-saeed-brainx** [comment]
Hi all, Any further update?

**Comment by osiset** [comment]
None yet. I've been super busy.

**Comment by emreisik95** [comment]
Is there any update?

**Comment by parkourben99** [comment]
Any further updates?

**Comment by niveshsaharan** [comment]
For someone interested in implementing JWT session token based authentication, I have implemented it in one of my own project. Here's the commit: https://github.com/niveshsaharan/shark/commit/1bd88f7bf91d0d30805cb9aa5c996bed3283fa55

**Comment by osiset** [comment]
No updates right now, I have been ill the past few weeks from a condition. I am trying my best to balance and make some time to tackle this frustrating browser issue. I am attempting to make time Sunday to account for this.

**Comment by darrynten** [comment]
#601 contains support for authentication via the JWT issued by Shopify.

**Comment by osiset** [comment]
Ugh. The loner Mac I have has the boot screen, with the logo, and just stays there. I'll be searching for a solution, but if anyone knows whats up with it off-by-hand, please let me know!
![PXL_20201101_232858181](https://user-images.githubusercontent.com/2420282/97818417-e77a3380-1c7c-11eb-8613-810b78874d71.jpg)

**Comment by myw-ben** [comment]
I am having issues that I think link in with this, but am not totally sure. 

A number of my existing apps are suddenly forcing users to the /login page. Oddly, its not all of them. Some pages work fine, others redirect.

Its definitely a cookie issue as it only happens on embedded apps, and not if it's just running as an external app. 

@osiset - I have an app in dev mode that I am happy to give you access to if you want to test out on it?

**Comment by arthurCreek** [comment]
I just want to mention a few things I've noticed, maybe it will help? 

I have tried to set 'secure' => env('SESSION_SECURE_COOKIE', true), and 'same_site' => 'none', but it doesn't help.  This is a solution I've seen that's worked for some people, but not for others.

Also, I've asked a few people if they use a Mac or windows laptop, so far non of the Mac users can replicate this issue.

**Comment by myw-ben** [comment]
@arthurCreek To chip in on this, I am on Mac and am facing these issues. Happens on Chrome & Safari, but not FF

**Comment by yrk2yrk** [comment]
I encountered the cookie issue on mac as well. I used Browserstack to check if it was specific to OS X version running on my mac, but that wasn't the case.

@osiset consider requesting a [BrowserStack open-source access](https://www.browserstack.com/open-source) to test this issue on mac. It works like a real thing.

**Comment by osiset** [comment]
@yrk2yrk I'll explore this, thanks.

**Comment by osiset** [comment]
I've been approved, thanks, will try this out.

**Comment by yrk2yrk** [comment]
That's awesome!

**Comment by osiset** [comment]
I'm still unable to reproduce :(

Standard settings...
http_only => true,
same_site => lax,
storage => file,

Tried with Safari on Big Sur, Chrome, Edge... app would load fine. However, inspecting the cookies in Safari for the app domain, everything would be blank, no cookies at all. So I'm not sure how its working?

**Comment by myw-ben** [comment]
It may be worth trying it in incognito mode. On some on my apps, it works in normal mode, but not in Incognito. 

This is affecting a few of my apps so I am more than happy to collaborate on this @osiset to see if we can sort it. 

Feel free to message me and I can give you access on one of my dev apps which is having this issue

**Comment by osiset** [comment]
So I am not sure what is going on or how its working.

Chrome incognito on Mac in BrowserStack - the homepage of my app fails to load, simply because my app is trying to access localStorage. However, inspecting the HTML source, I see the shop and other details, which means it passed through the middleware fine.

But, theres no cookies, yet user is marked as logged in.. I am not sure where its storing it :/

**Comment by osiset** [comment]
At this point, I am unsure what I need to do.

1. I've setup an ITP middleware to check for existence of a test cookie
2. If the test cookie does not exist, it does a fullpage redirect directly to the domain to a controller which sets the test cookie, this sets the cookie outside of the iframe
3. Redirect back to the app... stuck in a loop then

Going back to the app... request->cookies and $_COOKIE are empty for some reason.

Allowing third-party cookies, everything works fine, so I am unsure what I can do to get around this... any help would be greatful.

**Comment by osiset** [comment]
I'm wondering if the reason cookies are empty is because of the self-signed certificate. Even though I've added an exception, I wonder if its not sending the cookies due to this. I'll push up some stuff to a branch and deploy to a server to confirm.

**Comment by myw-ben** [comment]
@osiset I am running it on a production server using LetsEncrypt and still facing the issue. Not sure if this is self-signed or not ( or if it helps ), but wanted to throw it out there incase it helps

**Comment by osiset** [comment]
In your case no, the cookie setting outside the iframe is not part of the released versions.

**Comment by osiset** [comment]
I've tried a few more attempts both locally and on-server. No luck. I don't really understand why it is not working.

From my understanding, a samesite cookie needs to be set outside the iframe, which is done. I've even confirmed the cookie is there by navigating to a specific route and checking the cookie. However, the second I go back to the app within Shopify, no cookies are visible.

**Comment by nonick891** [comment]
Hey guys!

I am having the same issue like the `isFreemium() on null` but there is one exception it is happened in all browsers for me. Is it still the cookie issue?

The version of packages I am using:
`"osiset/basic-shopify-api": "^7.0"`
`"osiset/laravel-shopify": "^10.3.1"`

**Comment by osiset** [comment]
I have attempted this again this weekend without luck. I am unable to pinpoint what the issue is. Samesite cookie gets set outside the iframe, redirecting back to app, everything is empty for cookies.

**Comment by osiset** [comment]
I've had some success now by probing the Koa implementation.

I believe my flow needs to be this:

- ITP middleware will check for ITP cookie
- If ITP cookie does not exist, redirect to top level and set a same site cookie, then redirect back to app with a query param of ?itp=true
- Once redirected back to app, it'll run through ITP middleware again, we will check for ITP cookie, if it doesn't exist and the ITP query param exists, then we need to redirect to a cookie/storage permission page
- On this page, using JS, user will read a prompt and click to try and enable cookies / storage access
- Some browsers will prompt to confirm, others just allow it
- Once this is done, were good to redirect back and have cookie access


With this plan in mind and loosely working I'll continue to develop a solution.

**Comment by osiset** [comment]
I have success with the code, things seem to work OK. Test suite is passing (need to add test for new ITP middleware).

I just need to test Safari as well.

If anyone wants to try it out.. please see feature/itp branch.

Once I'm done the new test files, and confirm Safari, I'll make a release.

**Comment by osiset** [comment]
Hmm, things were working but now requestStorageAccess is not working properly... will look into it.

**Comment by osiset** [comment]
Now merged into master. Testers welcome. As noted in my PR...

>In my testing on Chrome and Edge, all is fine. Firefox and Safari, the requestStorageAccess is granted but cookie creation fails. There is a lot of docs about issues with Safari and requestStorageAccess, even on Shopify's own repos. For now, I can not find a work around, but a "Manual enable cookies" screen does display for those instances.

**Comment by osiset** [comment]
v16.0.0 released with those features. See changelog / upgrade log. New wiki entry [for ITP](https://github.com/osiset/laravel-shopify/wiki/ITP).

I will close any related issues for now, but I know this topic will come up again, Shopify's own repos have reports of this still not working 100% for them as well.

**Comment by amosmos** [comment]
GREAT WORK MAN!!!

**Comment by nguyenduc308** [comment]
I got the proplem when i navigated in shopify app using Safari (Chrome worked as well).  My app version using below:
```"laravel/framework": "^8.40",
"osiset/laravel-shopify": "^17.1"```
It was through error: Error Call to a member function getDomain() on null
Path: vendor/osiset/laravel-shopify/src/Traits/AuthController.php:32
============
Can anyone help me

**Comment by faizan-glowlogix** [comment]
@vladbova can you please share how did you solve this issue? Facing the same app is working fine on chrome but not working on safari. any help will be much appreciated.

**Comment by vladbova** [comment]
@faizan-glowlogix and @nguyenduc308 

I've managed to fix this.
Whenever you try to follow a link or click the app bridge button, always pass the 'shop' param into your route.

for example,

Instead of using this:
`{{ route('home') }}`

Use this:
`{{ route('home', ['shop' => auth()->user()->name ]) }}`


I hope it will help you too.

**Comment by faizan-glowlogix** [comment]
@vladbova thanks a lot. it worked.
